Allow multiple hostnames tags

aawsome opened the pull request:
What is the purpose of this change? What does it change?
--------------------------------------------------------

allows multiple hostnames for `dump`, `find`, `forget`, `ls`, `mount`,
`restore`, `snapshots`, `stats` and `tag`

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

closes #1570 

Checklist
---------

- [X] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [X] I have run `gofmt` on the code in all commits
- [X] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [X] I'm done, this Pull Request is ready for review

rawtaz commented on changelog/unreleased/issue-1570:
```diff
@@ -0,0 +1,5 @@
+Enhancement: Multiple hostnames for `dump`, `find`, `forget`, `ls`, `mount`, `restore`, `snapshots`, `stats`, `tag`.
+
+Added the possibility to specify multiple `--hostname` or `-H` flags where it makes sense.
```
The title line can't end with a dot, it shouldn't have any punctuation. I also suggest shortening it, please consider changing these three lines to the following (note that I also changed `--hostname` to `--host` since the former is deprecated):

~~~
Enhancement: Support specifying multiple host flags for various commands

Previously commands didn't take more than one `--host` or `-H` argument into account, which could be limiting with e.g. the `forget` command.

The `dump`, `find`, `forget`, `ls`, `mount`, `restore`, `snapshots`, `stats` and `tag` commands will now take into account multiple `--host` and `-H` flags.
~~~

rawtaz commented:
Nice work @aawsome! We need to adjust the changelog file a bit, once that's done I'll take a peek through the code again (just skimmed through it) and merge whenever possible.

fd0 commented on internal/restic/snapshot_find.go:
```diff
@@ -38,7 +38,12 @@ func FindLatestSnapshot(ctx context.Context, repo Repository, targets []string,
 		}
-		if snapshot.Time.Before(latest) || (hostname != "" && hostname != snapshot.Hostname) {
+
+		if snapshot.Time.Before(latest) {
```
Breaking up the complex condition makes the code so much easier to read, great idea!

fd0 commented on internal/restic/snapshot_find.go:
```diff
@@ -38,7 +38,12 @@ func FindLatestSnapshot(ctx context.Context, repo Repository, targets []string,
+			return nil
+		}
+
+		if hostnames != nil && !snapshot.HasHostname(hostnames) {
```
Good name for the function! If `HasHostname()` has defined behavior for `hostnames` being an empty (or nil) slice the code here becomes even better :)

aawsome replied:
I already thought about this and dismissed the idea for not being perfect consistent to what the name implies.. :smile: 
However, looking at `HasTagList()` there is already the same functionality. So I added this case, pointed it out in the description and also extended the description of `HasTagList()`. Now it is consistent that `Has"Object"` checks if one of the objects given in the list is satisfied and returns true for an empty list and `Has"Objects"` checks if all objects given in the list is satsified. IMO this makes sense.

fd0 commented on internal/restic/snapshot.go:
```diff
@@ -201,6 +201,17 @@ func (sn *Snapshot) HasPaths(paths []string) bool {
 }
 
+// HasHostname returns true if the snapshot hostname is in the list of the given hostnames.
+func (sn *Snapshot) HasHostname(hostnames []string) bool {
```
If `HasHostname()` accepts an empty slice and returns `true` in this case we can remove the `nil` checks below (needs to be mentioned in the description for the function though).

For slice emptyness, we should use `if len(hostnames) == 0`, this will match `hostnames == nil` as well as `hostnames == []string{}`.

aawsome replied:
is changed, see above.

fd0 commented on cmd/restic/cmd_ls.go:
```diff
@@ -84,7 +84,7 @@ type lsNode struct {
 
 func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
-	if len(args) == 0 && opts.Host == "" && len(opts.Tags) == 0 && len(opts.Paths) == 0 {
+	if len(args) == 0 && opts.Hosts == nil && len(opts.Tags) == 0 && len(opts.Paths) == 0 {
```
Use `len(opts.Hosts) == 0` instead of checking equality to `nil`, that's more robust.

aawsome replied:
is changed.

fd0 commented on cmd/restic/find.go:
```diff
@@ -39,7 +39,7 @@ func FindFilteredSnapshots(ctx context.Context, repo *repository.Repository, hos
 
 			// Give the user some indication their filters are not used.
-			if !usedFilter && (host != "" || len(tags) != 0 || len(paths) != 0) {
+			if !usedFilter && (hosts != nil || len(tags) != 0 || len(paths) != 0) {
```
Use `len(hosts) == 0` here, too

aawsome replied:
is changed.

fd0 commented on internal/restic/snapshot_find.go:
```diff
@@ -92,7 +97,7 @@ func FindFilteredSnapshots(ctx context.Context, repo Repository, host string, ta
 		}
 
-		if (host != "" && host != sn.Hostname) || !sn.HasTagList(tags) || !sn.HasPaths(paths) {
+		if (hosts != nil && !sn.HasHostname(hosts)) || !sn.HasTagList(tags) || !sn.HasPaths(paths) {
```
Same here, the check for `nil` is not needed (and not entirely correct) as well.

aawsome replied:
is changed.

fd0 reviewed (changes requested):
Awesome, thanks for your contribution! There's a few things I added comments to, in addition it'd be great if you could:
 1. Squash the commits
 2. Use a proper email address for the commits instead of `alex@thinkpad` (use `git config --global user.email` or something similar)

Thanks again!

aawsome committed:
Support specifying multiple host flags for various commands

The `dump`, `find`, `forget`, `ls`, `mount`, `restore`, `snapshots`,
`stats` and `tag` commands will now take into account multiple
`--host` and `-H` flags.

rawtaz reviewed (approved):
Approving this after @aawsome made changes requested by @fd0.

Edit: Apparently @fd0 has to approve it as well after having requested changes, I can't merge it as it is now. That's good!
